python/lsst/daf/butler/core/datastoreRecordData.py

```diff
@@ -82,15 +120,29 @@ class DatastoreRecordData:
     datastore.
     """
 
-    refs: List["DatasetIdRef"] = dataclasses.field(default_factory=list)
-    """List of DatasetRefs known to this datastore.
+    dataset_ids: Set[DatasetId] = dataclasses.field(default_factory=set)
+    """List of DatasetIds known to this datastore.
     """
 
-    records: Dict[str, List[StoredDatastoreItemInfo]] = dataclasses.field(
-        default_factory=lambda: defaultdict(list)
+    records: Dict[DatasetId, Dict[str, List[StoredDatastoreItemInfo]]] = dataclasses.field(
+        default_factory=lambda: defaultdict(lambda: defaultdict(list))
     )
-    """Opaque table data, grouped by opaque table name.
-    """
+    """Opaque table data, indexed by dataset ID and grouped by opaque table
+    name."""
+
+    def update(self, other: DatastoreRecordData) -> None:
+        """Update contents of this instance with data from another instance.
+
+        Parameters
+        ----------
+        other : `DatastoreRecordData`
+            Records tho merge into this instance.
+        """
+        self.dataset_ids.update(other.dataset_ids)
+        for dataset_id, table_records in other.records.items():
+            this_table_records = self.records.setdefault(dataset_id, {})
+            for table_name, records in table_records.items():
+                this_table_records.setdefault(table_name, []).extend(records)
```

Comment:
No need for `setdefault` here or two lines above since you're using `defaultdict`, and actually using `setdefault` with a plain `dict` will mess up other code that presumably does assume the nested thing is a `defaultdict`.  I think you need to stick to either `setdefault` or `defaultdict.__getitem__` across the board.